### ISSUE TITLE
Support premium search APIs

### ISSUE THREAD

**Issue by ddefranza** [tweepy/tweepy#1175]
With recent API updates, both basic and premium accounts have access to full-history search as long as a development environment is created and specified by the API account holder. Will tweepy be able to access this feature sometime in the near future?

**Comment by Dmium** [comment]
I'm not a contributor yet but after a glance at the code it looks like a fairly simple change to make it possible and a bit more to make it easy. I might make a PR for this if nobody else is

**Comment by Harmon758** [comment]
For reference:
https://developer.twitter.com/en/docs/tweets/search/overview/premium
https://twittercommunity.com/t/full-archive-search-now-available-as-a-premium-api/100718
https://blog.twitter.com/developer/en_us/topics/tools/2018/access-the-full-history-of-tweets.html

**Comment by JohnOmernik** [comment]
I am interested in the premium search APIs as well...

**Comment by ingmferrer** [comment]
Bump

**Comment by ejgutierrez74** [comment]
+1. Hope to see this feature soon....

**Comment by kcostya** [comment]
+1. I badly need this feature.

**Comment by Dmium** [comment]
I'll take a look tomorrow

**Comment by Dmium** [comment]
Update: got distracted but I have applied for new keys etc and am taking a look. If it takes more than a few hours though I might not finish this. It's a big project and I've not worked on it but it looks like it's structured sensibly enough to be a fairly simple change.

**Comment by akshaykhatter** [comment]
**Any updates for this issue?** (people have started moving to searchtweets and TwitterAPI but both seem to be in a fairly primitive stage and the documentation is not great.)

**Comment by feconroses** [comment]
+1 this feature would be super useful

**Comment by drandreaskrueger** [comment]
Oh, so there is an issue about this already.   
Yes, it would be great to use the Premium search from tweepy. 

I had started to just modify some existing code but got stuck at a `status_code=422` error, look   https://github.com/tweepy/tweepy/issues/1293

**Comment by Harmon758** [comment]
I've created a branch (https://github.com/tweepy/tweepy/tree/premium-search) and draft PR (#1294) that adds API methods for the data endpoints of both premium search APIs.
Cursor support with those methods is still pending. See #1294.

**Comment by drandreaskrueger** [comment]
Coooool, nice one, thanks!!

So how would I use this? I assume like that?

```
source env/bin/activate
pip uninstall tweepy
git clone https://github.com/tweepy/tweepy.git
cd tweepy
git checkout premium-search
python setup.py install
```
right? When I use `python setup.py install` it will end up in my virtualenv, as it would with `pip`, right? And what is the way to fully get rid of it again, if I want to get back to the current stable version 3.8.0 ?

Thanks!

**Comment by Harmon758** [comment]
You can simply `pip install --upgrade git+https://github.com/tweepy/tweepy@premium-search`.
See https://pip.pypa.io/en/stable/reference/pip_install/#git.
`pip install --force-reinstall --no-dependencies tweepy` will install the latest version from PyPI. You can remove the `--no-dependencies` option if you want to reinstall dependencies as well.
See https://pip.pypa.io/en/stable/reference/pip_install/#options.

**Comment by drandreaskrueger** [comment]
cool, thanks.

**Comment by MasterEric** [comment]
@Harmon758 Thank you for developing a pull request for this. The cursor changes required to continue development have been merged, and I would be happy to see continued progress.

**Comment by Harmon758** [comment]
I've added cursor support, and both methods should now be fully functional.
Feel free to try out and test them in [the branch](https://github.com/tweepy/tweepy/tree/premium-search) and [PR](https://github.com/tweepy/tweepy/pull/1294).

**Comment by smoufarrege** [comment]
Found what appear to be some bugs and missing features when using the search_30_day function in conjunction with paging. 

If you try to use a plus sign (+) in the query (i.e. for boolean searches), it crashes with the message "There were errors processing your request: no viable alternative at character '+'" -- I'm guessing a minus sign perhaps certain other characters would cause the same issue.

Also, my original algorithm used tweet_mode='extended' as one of the parameters. As per Twitter, this allows for a full 280 character tweet to be returned, rather than truncating it to the original 140 character limit. It looks like this isn't supported by the new function.

**Comment by Harmon758** [comment]
I'm not sure what you mean by "for boolean searches", as `+` isn't a valid operator for that (or any other, as far as I'm aware) API/endpoint.
See https://developer.twitter.com/en/docs/tweets/search/guides/premium-operators.
At any rate, this is an issue with Twitter's processing of the character (seemingly stemming from Salesforce/Apex) rather than with the Tweepy branch/PR.

As for extended Tweets, premium APIs are different from standard REST APIs and include an additional `extended_tweet` field in the payload, represented as an attribute for the corresponding Status object, exactly the same as Status objects from streams.
See https://developer.twitter.com/en/docs/tweets/tweet-updates and https://tweepy.readthedocs.io/en/latest/extended_tweets.html#streaming.

**Comment by eunbee31** [comment]
Hello, I would like to know what means `stale` on `premium-search` branch. Did it mean that the branch is not updated at all? That you will remove it eventually? 
Thanks in advance for your reply. Also, good work you have done :)

**Comment by Harmon758** [comment]
I'm not sure where you're seeing that.
The branch is not stale and should be fully functional.  I still plan to release it with Tweepy v3.9.
Once the PR is merged into the master branch, that branch will be deleted.

**Comment by eunbee31** [comment]
Thanks for your reply. I.m glad to hear that. 
I saw it here: https://github.com/tweepy/tweepy/branches

**Comment by Harmon758** [comment]
The Stale view on that page is simply GitHub showing all branches that no one has committed to in the last three months.
See https://docs.github.com/en/github/administering-a-repository/viewing-branches-in-your-repository.

**Comment by eunbee31** [comment]
Thanks.

**Comment by jikamens** [comment]
Did this make it into 3.9.0? I don't see it listed in the release notes.

**Comment by Harmon758** [comment]
v3.9.0 was released without #1294 having been merged yet. I had intended to release it with v3.9.0.
I've since fixed and improved the [changelog](https://github.com/tweepy/tweepy/blob/master/CHANGELOG.md#version-390) and [release notes](https://github.com/tweepy/tweepy/releases/tag/v3.9.0) for v3.9.0.
Since v3.9.0 was also released without originally noting that it would be the last version to support Python 2 (#1253) in its changelog and release notes, a v3.10.0 is now planned for that and this will be released as part of that version.
I'll be merging #1294 into master and deleting the `premium-search` branch shortly.

**Comment by stabilus** [comment]
just came across this and don't get how to include premium api support, since currently:
- pip install --upgrade git+https://github.com/tweepy/tweepy@premium-search is not working (as branch is not available)
- #1294 is not included in v3.9.0 yet
- v3.10. is not available yet

Any directions on how to proceed?

**Comment by Harmon758** [comment]
The branch has been deleted since #1294 has been merged into the master branch.
Until v3.10.0 is released, you can install from the master branch.

**Comment by stabilus** [comment]
> The branch has been deleted since #1294 has been merged into the master branch.
> Until v3.10.0 is released, you can install from the master branch.

ok. pip install --upgrade git+https://github.com/tweepy/tweepy@master successfully installed. 
But how do you now access premium search? In twitter dev, I created a new Dev environment and I have the corresponding label. How do you submit it to the API? How do you call premium search?

**Comment by Harmon758** [comment]
See the documentation for [`API.search_30_day`](https://tweepy.readthedocs.io/en/latest/api.html#API.search_30_day) and [`API.search_full_archive`](https://tweepy.readthedocs.io/en/latest/api.html#API.search_full_archive) and [the corresponding documentation for the premium search APIs](https://developer.twitter.com/en/docs/twitter-api/v1/tweets/search/api-reference/premium-search).

**Comment by stabilus** [comment]
ok. So, I'm not getting it. After reading just about the entire internet (including the twitter api documentation for the premium search APIs), I still don't have a working example on how I could translate my rather simple search string I'm using with the standard api ('(Bitcoin OR BTC OR XBT) -filter:retweets -filter:links' into a workable api.search_30_day( ... call.
An example would be so appreciated.

**Comment by andypiper** [comment]
Something like `query` as `(Bitcoin OR BTC OR XBT) -is:retweets -has:links` would be my starting point.

**Comment by imantha-das** [comment]
I managed to get accessibility to `search_full_archive` using the method stabilus described. But not sure exactly how to use `search_full_archive`.
Following some examples used `api.search`, I typed the following for `api_full_search` but **didn't** end up with an iterable object

`pastTweetsObj = tw.Cursor(
            api.search_full_archive,
            environment_name = 'VolcanicDisasterSearch',
            query = '#kilauea' +'-filter:retweets',
            fromDate = '201805010000',
            toDate ='201812310000'
        )`

I also tried removing the cursor object and typing as follows,

`pastTweetsObj = api.search_full_archive(
            environment_name = 'VolcanicDisasterSearch',
            query = '#kilauea' +'-filter:retweets',
            fromDate = '201805010000',
            toDate ='201812310000'
        )`

but then ended up with the following error. {'message': 'Forbidden: Authentication succeeded but account is not authorized to access this resource.', ...}

Any ideas how to use `api.search_full_archive` ?

**Comment by msklc** [comment]
Don't forget to update tweepy with `pip install --upgrade git+https://github.com/tweepy/tweepy@master`. Then you can use the API with premium feature...

**Comment by Harmon758** [comment]
@imantha-das Make sure your [dev environment](https://developer.twitter.com/en/account/environments) is properly set up.

@stabilus @imantha-das This is not pertinent to this issue. Consider [starting a separate discussion](https://github.com/tweepy/tweepy/discussions/new) or joining the [Discord server](https://discord.gg/bJvqnhg), also linked in the README, if you need further help.

**Comment by amillb** [comment]
It seems that the premium search is still missing from the standard pip and conda installs with 3.10.0. The `pip install --upgrade git+https://github.com/tweepy/tweepy@master` method still works, but is there a reason that the `Client` class is missing from the version on pip?

**Comment by Harmon758** [comment]
The premium search APIs were implemented with [`API.search_30_day`](https://tweepy.readthedocs.io/en/latest/api.html#tweepy.API.search_30_day) and [`API.search_full_archive`](https://tweepy.readthedocs.io/en/latest/api.html#tweepy.API.search_full_archive) and released as part of v3.10.0.
`Client` is used to interface with Twitter API v2, not the premium search APIs, and is expected to be released with v4.0.

**Comment by amillb** [comment]
Aha, thanks for clarifying. I saw `Client` in the [docs](https://docs.tweepy.org/en/latest/client.html), and thought that it had already been released.

**Comment by Harmon758** [comment]
The `latest` version of the docs is for the development version on the master branch (why it says 4.0.0-alpha in the title and Revision ef810334 in the footer). You can refer to the `v3.10.0` version of the docs, https://tweepy.readthedocs.io/en/v3.10.0/, for v3.10.0.

**Comment by amillb** [comment]
Oh, thanks for clarifying, The docs came up in a google search and I didn't notice the version. Apologies for the bother.
